File: app/src/main/java/com/example/vakildisk/repository/FirestoreRepository.java
package com.example.vakildisk.repository;

import com.example.vakildisk.models.Case;
import com.example.vakildisk.models.Hearing;
import com.example.vakildisk.models.User;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.firebase.Timestamp;
import com.google.firebase.firestore.CollectionReference;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.EventListener;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.Query;
import com.google.firebase.firestore.QuerySnapshot;

import java.util.HashMap;
import java.util.Map;

public class FirestoreRepository {

    private final FirebaseFirestore db;
    private final CollectionReference usersRef;
    private final CollectionReference casesRef;
    private final CollectionReference hearingsRef;

    public FirestoreRepository() {
        this.db = FirebaseFirestore.getInstance();
        this.usersRef = db.collection("users");
        this.casesRef = db.collection("cases");
        this.hearingsRef = db.collection("hearings");
    }

    // --- User Operations ---

    public void createUser(User user, OnCompleteListener<Void> listener) {
        usersRef.document(user.getUserId()).set(user).addOnCompleteListener(listener);
    }

    public void getUserById(String userId, OnCompleteListener<com.google.firebase.firestore.DocumentSnapshot> listener) {
        usersRef.document(userId).get().addOnCompleteListener(listener);
    }

    // --- Case Operations ---

    public void createCase(Case caseObj, OnCompleteListener<DocumentReference> listener) {
        casesRef.add(caseObj).addOnCompleteListener(task -> {
            if (task.isSuccessful() && task.getResult() != null) {
                String id = task.getResult().getId();
                casesRef.document(id).update("caseId", id);
            }
            listener.onComplete(task);
        });
    }

    public void updateCase(String caseId, Map<String, Object> updates, OnCompleteListener<Void> listener) {
        casesRef.document(caseId).update(updates).addOnCompleteListener(listener);
    }

    public void getCasesBySeniorId(String seniorId, EventListener<QuerySnapshot> listener) {
        casesRef.whereEqualTo("seniorId", seniorId)
                .orderBy("createdAt", Query.Direction.DESCENDING)
                .addSnapshotListener(listener);
    }

    public void getCasesByJuniorId(String juniorId, EventListener<QuerySnapshot> listener) {
        casesRef.whereEqualTo("juniorId", juniorId)
                .orderBy("createdAt", Query.Direction.DESCENDING)
                .addSnapshotListener(listener);
    }

    // --- Hearing Operations ---

    public void createHearing(Hearing hearing, OnCompleteListener<DocumentReference> listener) {
        hearingsRef.add(hearing).addOnCompleteListener(task -> {
            if (task.isSuccessful() && task.getResult() != null) {
                String id = task.getResult().getId();
                hearingsRef.document(id).update("hearingId", id);
                
                // When a hearing is created, update the nextHearingDate in the case document
                Map<String, Object> update = new HashMap<>();
                update.put("nextHearingDate", hearing.getHearingDate());
                updateCase(hearing.getCaseId(), update, task1 -> {});
            }
            listener.onComplete(task);
        });
    }

    public void updateHearingDate(String caseId, Timestamp newDate, OnCompleteListener<Void> listener) {
        Map<String, Object> update = new HashMap<>();
        update.put("nextHearingDate", newDate);
        casesRef.document(caseId).update(update).addOnCompleteListener(listener);
    }

    public void getHearingsByCaseId(String caseId, EventListener<QuerySnapshot> listener) {
        hearingsRef.whereEqualTo("caseId", caseId)
                .orderBy("hearingDate", Query.Direction.ASCENDING)
                .addSnapshotListener(listener);
    }
}

File: app/src/main/java/com/example/gpt/FirestoreRepository.kt
package com.example.gpt

import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import com.google.firebase.firestore.FieldValue
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.tasks.await

class FirestoreRepository(private val db: FirebaseFirestore) {

    suspend fun createUser(user: UserModel): Result<Void?> {
        return try {
            db.collection("users").document(user.id).set(user).await()
            Result.success(null)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun getUser(userId: String): Result<UserModel?> {
        return try {
            val snapshot = db.collection("users").document(userId).get().await()
            Result.success(snapshot.toObject(UserModel::class.java))
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun updateFcmToken(userId: String, token: String): Result<Void?> {
        return try {
            db.collection("users").document(userId).update("fcmToken", token).await()
            Result.success(null)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun createCase(case: CaseModel): Result<String> {
        return try {
            val ref = db.collection("cases").document()
            val newCase = case.copy(id = ref.id)
            ref.set(newCase).await()
            Result.success(ref.id)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    fun getCasesForSenior(seniorId: String): Flow<List<CaseModel>> = callbackFlow {
        val subscription = db.collection("cases")
            .whereEqualTo("seniorLawyerUID", seniorId)
            .orderBy("createdAt", Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, error ->
                if (error != null) return@addSnapshotListener
                if (snapshot != null) {
                    trySend(snapshot.toObjects(CaseModel::class.java))
                }
            }
        awaitClose { subscription.remove() }
    }

    fun getCasesForJunior(juniorId: String): Flow<List<CaseModel>> = callbackFlow {
        val subscription = db.collection("cases")
            .whereArrayContains("juniorLawyerUIDs", juniorId)
            .orderBy("createdAt", Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, error ->
                if (error != null) return@addSnapshotListener
                if (snapshot != null) {
                    trySend(snapshot.toObjects(CaseModel::class.java))
                }
            }
        awaitClose { subscription.remove() }
    }

    suspend fun addHearing(hearing: HearingModel): Result<Void?> {
        return try {
            val ref = db.collection("hearings").document()
            val newHearing = hearing.copy(id = ref.id)
            db.runBatch { batch ->
                batch.set(ref, newHearing)
                // Also update the nextHearingDate in the Case document
                val caseRef = db.collection("cases").document(hearing.caseId)
                batch.update(caseRef, "nextHearingDate", hearing.date)
            }.await()
            Result.success(null)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun assignJuniorToCase(caseId: String, juniorId: String): Result<Void?> {
        return try {
            db.collection("cases").document(caseId)
                .update("juniorLawyerUIDs", FieldValue.arrayUnion(juniorId))
                .await()
            Result.success(null)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}